### PR TITLE
P1523R1 Views and Size Types

### LINKED ISSUE #3034: [2019-07 LWG Motion 30] P1523R1 Views and Size Types

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #3034.

### CODE REVIEW

**Comment by burblebee** [comment]
I used \hfull\break to work around an Overfull issue - what should I have used?

Also, shouldn't make-unsigned-like be introduced and explained somewhere?  It first appears out of the blue in [range.prim.size]p1.3 (same for make-unsigned-like-t).

**Comment by jwakely** [comment]
They're introduced in P1522R1.

File: source/ranges.tex [hunk 1]
```diff
@@ -522,10 +522,10 @@
   \end{itemize}
 
 \item
-  Otherwise, \tcode{(ranges::end(E) - ranges::begin(E))}
+  Otherwise, \tcode{\placeholder{make-unsigned-like}(ranges::end(E) - ranges::begin(E))}\iref{range.subrange}
```
Comment (jensmaurer) [inline source/ranges.tex]: Should this be \placeholdernc instead?

File: source/ranges.tex [hunk 2]
```diff
@@ -1151,18 +1151,18 @@
...
-    static constexpr bool StoreSize =                   // \expos
+    static constexpr bool StoreSize =                      // \expos
       K == subrange_kind::sized && !SizedSentinel<S, I>;
-    I begin_ = I();                                     // \expos
-    S end_ = S();                                       // \expos
-    iter_difference_t<I> size_ = 0;                     // \expos; present only
-                                                        // when \tcode{StoreSize} is \tcode{true}
+    I begin_ = I();                                        // \expos
+    S end_ = S();                                          // \expos
+    @\placeholder{make-unsigned-like-t}@(iter_difference_t<I>) size_ = 0;  // \expos; present only
```
Comment (jensmaurer) [inline source/ranges.tex]: \placeholdernc ?

File: source/ranges.tex (hunk 1 above)
Comment (burblebee) [reply source/ranges.tex]: Should it?  I've seen both used within tcode contexts throughout the spec; it's not clear when to use which - can you/someone clarify?

Comment (johelegp) [reply source/ranges.tex]: From what I've seen, one's used within `\tcode`s and the other within code blocks.

Comment (zygoloid) [reply source/ranges.tex]: Use the nc form to turn off italic correction in places where it looks bad. I think we have some examples in the style guide.

Comment (zygoloid) [reply source/ranges.tex]: Looks like we don't have anything in the style guide for this. We should :)

Comment (burblebee) [reply source/ranges.tex]: So what should be used in this case?  And yes, an example in the style guide would be very helpful.

File: source/ranges.tex (hunk 2 above)
Comment (burblebee) [reply source/ranges.tex]: ... and what should be used in this case?

File: source/ranges.tex (hunk 1 above)
Comment (jensmaurer) [reply source/ranges.tex]: My personal rule-of-thumb is to use \placeholdernc if there's something immediately afterwards (no space), and use \placeholder if a space follows.

Comment (burblebee) [reply source/ranges.tex]: ok, done.

File: source/ranges.tex (hunk 2 above)
Comment (burblebee) [reply source/ranges.tex]: changed also.

File: source/ranges.tex [hunk 3]
```diff
@@ -1151,18 +1151,18 @@
...
+    static constexpr bool StoreSize =                      // \expos
       K == subrange_kind::sized && !SizedSentinel<S, I>;
-    I begin_ = I();                                     // \expos
-    S end_ = S();                                       // \expos
-    iter_difference_t<I> size_ = 0;                     // \expos; present only
-                                                        // when \tcode{StoreSize} is \tcode{true}
+    I begin_ = I();                                        // \expos
+    S end_ = S();                                          // \expos
+    @\placeholdernc{make-unsigned-like-t}@(iter_difference_t<I>) size_ = 0;  // \expos; present only
+                                                           // when \tcode{StoreSize} is \tcode{true}
```
Comment (jensmaurer) [inline source/ranges.tex]: Can these go on a single line?

Comment (burblebee) [reply source/ranges.tex]: done.

Comment (burblebee) [reply source/ranges.tex]: @jensmaurer can you approve now?

**Comment by burblebee** [comment]
> Most of your uses of \placeholder in this change are similar to the uses where you changed to \placeholdernc. Did you make up your mind what looks better? Then, please be consistent.

I changed all added \placeholder{}s with \placeholdernc{}s to make you happy, but the surrounding Latex is inconsistent in its use of \placeholder v. \placeholdernc.  We need to come up with some rules on which to use where, and do a global sweep to fix  it.  

Leave this unresolved so we'll know to create an issue for it.

**Comment by zygoloid** [comment]
> > Most of your uses of \placeholder in this change are similar to the uses where you changed to \placeholdernc. Did you make up your mind what looks better? Then, please be consistent.
> 
> I changed all added \placeholder{}s with \placeholdernc{}s to make you happy, but the surrounding Latex is inconsistent in its use of \placeholder v. \placeholdernc. We need to come up with some rules on which to use where, and do a global sweep to fix it.

Please don't blindly change `\placeholder` to `\placeholdernc`. The latter is a surgical tool for fixing cases of bad kerning due to italics corrections, and should be used sparingly only where there is actual bad kerning. Usually the italics correction is what we want; the bounding box for italicized text is wider due to the right overhang, and it so we want to create a little extra space.

I've found that `\placeholdernc` improves the presentation immediately before a period or a comma (where there is no higher text for the italics to get uncomfortably close to). So let's take that as rough guidance for now: use `\placeholdernc` before a teletype period or comma, where there would otherwise be an unsightly amount of horizontal space.

**Comment by zygoloid** [comment]
Looking through a few more examples, `\placeholdernc` before a left paren also seems more pleasant to my eye, but before a left angle bracket I prefer including the italic correction.

**Comment by burblebee** [comment]
> > Most of your uses of \placeholder in this change are similar to the uses where you changed to \placeholdernc. Did you make up your mind what looks better? Then, please be consistent.
> 
> I changed all added \placeholder{}s with \placeholdernc{}s to make you happy, but the surrounding Latex is inconsistent in its use of \placeholder v. \placeholdernc. We need to come up with some rules on which to use where, and do a global sweep to fix it.
> 
> Leave this unresolved so we'll know to create an issue for it.

opened #3135

**Comment by burblebee** [comment]
> Looking through a few more examples, `\placeholdernc` before a left paren also seems more pleasant to my eye, but before a left angle bracket I prefer including the italic correction.

All the \placeholder{}s added in this motion occur before left parens, so it appears that the changes from \placeholder -> \placeholdernc in the Fixups are correct.

File: source/ranges.tex (hunk 1 above)
Comment (johelegp) [reply source/ranges.tex]: > From what I've seen, one's used within `\tcode`s and the other within code blocks.

I was confused. The difference I actually noticed was that one had to be wrapped in `@`s and the other not when in `\tcode`s or not, not whether it had to be `\placeholdernc` or the other.

**Pull request merged**